# Genome Assembly

## Environment preparation

Dependency:

- [Conda](https://docs.conda.io/en/latest/miniconda.html)
- [Anvi'o 7.1](https://anvio.org/install/)

### If you use BioCluster

You can simply load this environment in Shell:

```sh
conda activate anvio-7.1
export PATH=/export/home/CFIA-ACIA/chuanj/software/bin:$PATH
```

If this fails, use:

```sh
source /export/home/CFIA-ACIA/chuanj/.rc
conda_init
conda activate anvio-7.1
cjc_exports
```

Working directory:

```sh
DIR=/isilon/ottawa-rdc/users/chenw_lab/Eric/LinuxTraining/GenomeAssembly

## Display all files under $DIR
find $DIR -type f
# ./fastq/S74_R1_001.atria.fastq.gz
# ./fastq/S74_R2_001.atria.fastq.gz
# ./fastq/S1_R1_001.atria.fastq.gz
# ./fastq/S1_R2_001.atria.fastq.gz
```

We have two paired-end Illumina sequence samples: S74 and S1.

R1 and R2 means the Illumina fastq files are paired. R1 is forward (read 1), and R2 is reverse (read 2).

## Run Assembly

Here we introduce an assembly software, `megahit`.

### Megahit de novo assembly program

To get the command help of `megahit`, just type `megahit -h`. I list some of the useful help below:

```
MEGAHIT v1.2.9

contact: Dinghua Li <voutcn@gmail.com>

Usage:
  megahit [options] {-1 <pe1> -2 <pe2> | --12 <pe12> | -r <se>} [-o <out_dir>]

  Input options that can be specified for multiple times (supporting plain text and gz/bz2 extensions)
    -1                       <pe1>          comma-separated list of fasta/q paired-end #1 files, paired with files in <pe2>
    -2                       <pe2>          comma-separated list of fasta/q paired-end #2 files, paired with files in <pe1>

...

    -t/--num-cpu-threads     <int>          number of CPU threads [# of logical processors]

...

    -o/--out-dir             <string>       output directory [./megahit_out]
    --out-prefix             <string>       output prefix (the contig file will be OUT_DIR/OUT_PREFIX.contigs.fa)
...

    -h/--help                               print the usage message
    -v/--version                            print version
```

According to the help description, how do we analyze the samples to a specific output directory? Assume the output directory is `$DIR/$(whoami)`

```sh
## change directory
cd $DIR


## Define variables
READ1=$DIR/fastq/S74_R1_001.atria.fastq.gz
READ2=$DIR/fastq/S74_R2_001.atria.fastq.gz
SAMPLE=$( basename $READ1 | cut -f 1 -d _ )  # print the basename of READ1; then cut the basename by delimiter (underscore), and return the 1st field
OUTDIR=$DIR/$(whoami)/megahit_$SAMPLE  # whoami is a command to print your username

## You can double check variables by printing them
echo $READ1
# /isilon/ottawa-rdc/users/chenw_lab/Eric/LinuxTraining/GenomeAssembly/fastq/S74_R1_001.atria.fastq.gz
echo $READ2
# /isilon/ottawa-rdc/users/chenw_lab/Eric/LinuxTraining/GenomeAssembly/fastq/S74_R2_001.atria.fastq.gz
echo $OUTDIR
# /isilon/ottawa-rdc/users/chenw_lab/Eric/LinuxTraining/GenomeAssembly/**
echo $OUTPREFIX
# S74

# make directory
mkdir -p `dirname $OUTDIR`

## Main magahit command
megahit -1 $READ1 -2 $READ2 -o $OUTDIR -t 32

## Output layout
ls -ahl $OUTDIR
# total 9.4M
# drwxrwxr-x 3 ** ** users  178 Mar 30 15:29 .
# drwxrwxr-x 3 ** ** users   29 Mar 30 15:28 ..
# -rwxrwxr-x 1 ** ** users  262 Mar 30 15:29 checkpoints.txt
# -rwxrwxr-x 1 ** ** users    0 Mar 30 15:29 done
# -rwxrwxr-x 1 ** ** users 6.4M Mar 30 15:29 final.contigs.fa
# drwxrwxr-x 2 ** ** users 2.8K Mar 30 15:29 intermediate_contigs
# -rwxrwxr-x 1 ** ** users 141K Mar 30 15:29 log
# -rwxrwxr-x 1 ** ** users 1.2K Mar 30 15:28 options.json
```

Your assembled contig file is `$OUTDIR/megahit_$SAMPLE/final.contigs.fa`

Display the contig file using `less $OUTDIR/megahit_$SAMPLE/final.contigs.fa`

```
>k141_48 flag=0 multi=47.0000 len=218
CGTTTTATCCGGCCGTTCTCGAGTTCGAAGGAATGGATGTAGTGGTTCTCGCAATAGCCCTGCGGATACCCCGGGACCAGCGCCTTGCCGCGCCCGTCGCACTCGACCCAGAAGTGGTTCGGATCCTCGGTCTCGAAGATCCGCACGTTGTGCCACTCCCAGTCGGGGAAGCAGCGCTCGAGCCACTCGGCGAGCCGCCTGAGGCTCTCATGGCCCCG
>k141_72 flag=1 multi=1.0000 len=276
AGTCTCATTCGGAGCGAGCCCCGCTGATGTTTAACTCGACCTGCCGGAATCGCCCGGCCCCACCAGAGAGGATCGACCCCATGTGGACCAAGCCCAGCTTCACCGACCTGCGTCTCGGTTTCGAAGTGACCCTCTACTTCTTCAACCAGGGCGAGGTGTGCACCTGCCCGTCCCGCGCGCTGGTCCAGGAGTCGATCTACCCGGCGTTCATGGAAGAGGTGCTGAAAAAGGTCAGGGCGATCAAGCGCGGCGACCCGCTGGACACCGAGACCATGG
```

### Run samples in batch (for loop)

```sh
## change directory
cd $DIR

for READ1 in $DIR/fastq/*_R1_001.atria.fastq.gz
do
    echo Starting: $READ1

    local READ2=${READ1/_R1_001/_R2_001}
    local SAMPLE=$( basename $READ1 | cut -f 1 -d _ ) 
    local OUTDIR=$DIR/$(whoami)/megahit_$SAMPLE

    mkdir -p `dirname $OUTDIR`

    megahit -1 $READ1 -2 $READ2 -o $OUTDIR -t 32
done
```

### Quality assessment of assembled contigs

We need to check the assembly quality of contigs.

Contig files: `$DIR/$(whoami)/megahit_*/final.contigs.fa`

Program: `quast` or `checkm`.

Here I use `quast` as an example:

```sh
quast --threads 10 --output-dir $DIR/$(whoami)/qc_quast_results $DIR/$(whoami)/megahit_*/final.contigs.fa
```

Outputs:

```
Main parameters: 
  MODE: default, threads: 10, minimum contig length: 500, minimum alignment length: 65, \
  ambiguity: one, threshold for extensive misassembly size: 1000

RESULTS:
  Text versions of total report are saved to /isilon/ottawa-rdc/users/chenw_lab/Eric/LinuxTraining/GenomeAssembly/CFIA-ACIA+chuanj/qc_quast_results/report.txt, report.tsv, and report.tex
  Text versions of transposed total report are saved to /isilon/ottawa-rdc/users/chenw_lab/Eric/LinuxTraining/GenomeAssembly/CFIA-ACIA+chuanj/qc_quast_results/transposed_report.txt, transposed_report.tsv, and transposed_report.tex
  HTML version (interactive tables and plots) is saved to /isilon/ottawa-rdc/users/chenw_lab/Eric/LinuxTraining/GenomeAssembly/CFIA-ACIA+chuanj/qc_quast_results/report.html
  PDF version (tables and plots) is saved to /isilon/ottawa-rdc/users/chenw_lab/Eric/LinuxTraining/GenomeAssembly/CFIA-ACIA+chuanj/qc_quast_results/report.pdf
  Icarus (contig browser) is saved to /isilon/ottawa-rdc/users/chenw_lab/Eric/LinuxTraining/GenomeAssembly/CFIA-ACIA+chuanj/qc_quast_results/icarus.html
  Log is saved to /isilon/ottawa-rdc/users/chenw_lab/Eric/LinuxTraining/GenomeAssembly/CFIA-ACIA+chuanj/qc_quast_results/quast.log
```

Browse the text version of report:

```
less /isilon/ottawa-rdc/users/chenw_lab/Eric/LinuxTraining/GenomeAssembly/CFIA-ACIA+chuanj/qc_quast_results/report.txt
```

Report output:
```
All statistics are based on contigs of size >= 500 bp, unless otherwise noted (e.g., "# contigs (>= 0 bp)" and "Total length (>= 0 bp)" include all contigs).

Assembly                    megahit_S1_final.contigs  megahit_S74_final.contigs
# contigs (>= 0 bp)         32                        107                      
# contigs (>= 1000 bp)      15                        77                       
# contigs (>= 5000 bp)      11                        58                       
# contigs (>= 10000 bp)     11                        50                       
# contigs (>= 25000 bp)     11                        45                       
# contigs (>= 50000 bp)     8                         32                       
Total length (>= 0 bp)      4,006,832                 6,674,835                  
Total length (>= 1000 bp)   4,000,548                 6,661,438                  
Total length (>= 5000 bp)   3,991,826                 6,620,217                  
Total length (>= 10000 bp)  3,991,826                 6,567,041                  
Total length (>= 25000 bp)  3,991,826                 6,493,940                  
Total length (>= 50000 bp)  3,879,667                 6,026,871                  
# contigs                   18                        84                       
Largest contig              2,301,219                 453,651                   
Total length                4,002,329                 6,667,035                  
GC (%)                      38.48                     66.09                    
N50                         2,301,219                 252,074                   
N75                         374,255                   106,461                   
L50                         1                         10                       
L75                         3                         20                       
# N's per 100 kbp           0.00                      0.00                     
```

### Report explanation:

Reference: [QUAST 5.2.0 Manual Section 3.1.1](https://quast.sourceforge.net/docs/manual.html#sec3.1.1)

**\# contigs (≥ x bp)** is total number of contigs of length ≥ x bp. Not affected by the --min-contig parameter (see section 2.3).

**Total length (≥ x bp)** is the total number of bases in contigs of length ≥ x bp. Not affected by the --min-contig parameter (see section 2.3).

All remaining metrics are computed for contigs that exceed the threshold specified with --min-contig (see section 2.3, default is 500 bp).

**\# contigs** is the total number of contigs in the assembly.

**Largest contig** is the length of the longest contig in the assembly.

**Total length** is the total number of bases in the assembly.

**Reference length** is the total number of bases in the reference genome.

**GC (%)** is the total number of G and C nucleotides in the assembly, divided by the total length of the assembly.

**Reference GC (%)** is the percentage of G and C nucleotides in the reference genome.

**N50** is the length for which the collection of all contigs of that length or longer covers at least half an assembly.
NG50 is the length for which the collection of all contigs of that length or longer covers at least half the reference genome.
This metric is computed only if the reference genome is provided.

**Nx and NGx (for x between 0 and 100)** are defined similarly to N50 but with x % instead of 50 %. The value of  x  is set with --x-for-Nx (90 by default).

**L50 (Lx, LG50, LGx)** is the number of contigs equal to or longer than N50 (Nx, NG50, NGx)
In other words, L50, for example, is the minimal number of contigs that cover half the assembly.

# Try yourself

## Alternative de novo assembly program: MaSuRCA

MaSuRCA is a program combining assembly, error correction and scaffolding. 

Input can be Illumina sequences and long sequences from Nanopore and PacBio. You can also perform hybrid assembly if you have both short (Illumina) and long (Nanopore/PacBio) reads.

Typing `masurca -h` shows the help page of MaSuRCA:

```
masurca -r paired_ends_fwd.fastq.gz,paired_ends_rev.fastq.gz -t 32

Options:
 -t, --threads             ONLY to use with -i option, number of threads
 -i, --illumina            Run assembly without creating configuration file, argument can be 
                              illumina_paired_end_forward_reads 
                                or 
                              illumina_paired_end_forward_reads,illumina_paired_end_reverse_reads
                           if you only have single-end Illumina data. Reads file(s) could be fasta or fastq, can be gzipped.
 -r, --reads               ONLY to use with -i option, single long reads file for hybrid assembly, can be Nanopore or PacBio, 
                           fasta or fastq, can be gzipped

 -v, --version             Report version
 -o, --output              Assembly script (assemble.sh)
 -g, --generate            Generate example configuration file
 -p, --path                Prepend to PATH in assembly script
 -l, --ld-library-path     Prepend to LD_LIBRARY_PATH in assembly script
     --skip-checking       Skip checking availability of other executables
 -h, --help                This message
```

Like the previous script, the following command can run the MaSuRCA assembly:

```sh
## change directory
cd $DIR

## Define variables
READ1=$DIR/fastq/S74_R1_001.atria.fastq.gz
READ2=$DIR/fastq/S74_R2_001.atria.fastq.gz
SAMPLE=$( basename $READ1 | cut -f 1 -d _ )  # print the basename of READ1; then cut the basename by delimiter (underscore), and return the 1st field
OUTDIR=$DIR/$(whoami)/masurca_$SAMPLE  # whoami is a command to print your username

# make directory
mkdir -p $OUTDIR

# IMPORTANT!!! Have to change to the directory and run masurca
cd $OUTDIR

## Main magahit command
masurca -i $READ1,$READ2 -o assemble.sh -t 1

## Output layout
ls -ahl $OUTDIR
```